Added edi for 1476E (#6335)
cpinitiative/usaco-guide: solutions/gold/cf-1476E.mdx

---
id: cf-1476E
source: CF
title: Pattern Matching
author: Abhinav Singh
---

[Official Analysis (C++)](https://codeforces.com/blog/entry/87356)

## Explanation

> We are given patterns of length $k$ containing lowercase letters and `_`,which are just positions which can become **any** letter, and queries consisting of strings with a required pattern index. We realise a string matches a pattern if all the characters at the non-`_` indices are the same.

Reviewer: ```suggestion
For a string to match a pattern, all non-`_` indices in the pattern must match with the string.
```
I like this short and sweet version a lot more, as people knows what they're dealing with in the problem. Feel free to find a middle ground though.

Author: the reason i had this line was cuz the question was worded REALLY tough like even i had to read the question a few times before understanding what it was asking so i just restated it in the first part

Author: again i'd love someone else's suggestion on this one.... (looking at you @Sosuke23)

Reviewer: @4nav In this case I think you can just restate what the queries are in a sentence only, and then use my suggestion text to explain how matching works in another clause..  would like another opinion on whether it’s needed at all but you can definitely shorten the current form

We want to find an ordering of the patterns such that for each query $(s, p)$, the pattern $p$ is the **first** match for $s$ in that ordering, or every other pattern in $S$ (the set of all patterns matching $s$) must appear after $p$.

> For each query $(s,p)$, let $S$ be the set of all matching patterns and let $p$ be the required one. We are told to find an ordering of patterns such that for every query, $p$ appears before every other pattern in $S$. Each requirement is just a **precedence constraint**, i.e $p$ must come before $x$, that we model as a directed edge $p \to x$.

Reviewer: There's lots of repetition here.  Something like this would probably work better if I understood you correctly:

> Let $S$ be the set of all patterns matching $s$.
Thus, for each query $(s, p)$, the pattern $p$ is the **first** element in $S$ in the ordering, and all further elements of $S$ must appear after $p$.
Each requirement is therefore a **precedence constraint**, where $p$ cannot come before any other elements in $S$.
We model these as directed edges, $p \to x$, where $x$ is any non-$p$ element in $S$.

Again since I didn't even make Gold I'm dumb so there's prob some better way to word this.  orz orz orz

```suggestion
Let $S$ be the set of all patterns matching $s$.
Thus, for each query $(s, p)$, the pattern $p$ is the **first** element in $S$ in the ordering, and all further elements of $S$ must appear after $p$.
Each requirement is therefore a **precedence constraint**, where $p$ cannot come before any other elements in $S$.
We model these as directed edges, $p \to x$, where $x$ is any non-$p$ element in $S$.
```

Author: oh i mixed 2 drafts into the final lol sry will fix

A valid ordering then corresponds to a topological sort of the resulting graph: if the constraints are consistent, any topological order works; if two constraints contradict each other (e.g. $p \to x$ and $x \to p$), a cycle forms and no valid ordering exists.

To compute $S$, note that since $k \le 4$, each query string has at most $2^k \le 16$ wildcard variants, obtained by replacing subsets of positions with `_`.

> We store a hashmap from pattern string to index and look up each variant to collect all matching patterns efficiently.
> 
> For each query:
> 1. Generate $S$ via the $2^k$ wildcard variants.
> 2. If $p \notin S$, immediately output `NO`.
> 3. Otherwise, add directed edges $p \to x$ for all $x \in S$, $x \neq p$.
> After processing all queries, run a topological sort on the pattern graph. If a cycle exists, print `NO`; otherwise, output the topological order.

Reviewer: ```suggestion
For each query:

1. Generate $S$ by enumerating all $2^k$ wildcard variants.
2. If $p \notin S$, immediately output `NO`.
3. Otherwise, add directed edges $p \to x$ for all $x \in S$, $x \neq p$.

After processing all queries, run a topological sort on the pattern graph. If a cycle exists, print `NO`; otherwise, output the topological order.
```

I think spaces are needed to render properly, also 1 wording nit

## Implementation
Since $k \le 4$, each query generates at most $2^k = 16$ variants, so building the graph takes $\mathcal{O}(m \cdot 2^k)$ time. The topological sort runs in
$\mathcal{O}(n + m \cdot 2^k)$, giving an overall complexity of $\mathcal{O}(n + m \cdot 2^k)$.

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> adj;
vector<int> color;  // 0 = unvisited, 1 = in current DFS path, 2 = fully processed
vector<int> topo_order;

// Returns false if a cycle is detected
bool dfs(int u) {
	color[u] = 1;
	for (int v : adj[u]) {
		if (color[v] == 1) return false;  // back edge = cycle
		if (color[v] == 0) {
			bool no_cycle = dfs(v);
			if (!no_cycle) return false;
		}
	}
	color[u] = 2;
	topo_order.push_back(u);
	return true;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int n, m, k;
	cin >> n >> m >> k;

	// Map from pattern string to its 0-indexed position in input
	unordered_map<string, int> pat_idx;
	for (int i = 0; i < n; i++) {
		string s;
		cin >> s;
		pat_idx[s] = i;
	}

	vector<pair<string, int>> queries(m);
	vector<vector<int>> matches(

> 	    m);  // matches[i] = indices of all patterns that match query i

Reviewer: Formatting is funky here... move comment to previous line?

---
id: cf-1476E
source: CF
title: Pattern Matching
author: Abhinav Singh
---

[Official Analysis (C++)](https://codeforces.com/blog/entry/87356)

## Explanation

> We are given patterns of length $k$ containing lowercase letters and `_`, and queries consisting of strings with a required pattern index. For a string to match a pattern, all non-`_` indices in the pattern must match with the string.

Reviewer 2: This is more so just a thought but idk I find noting each string as query a little weird but its ok ig.

For each query $(s, p)$, let $S$ be the set of all patterns matching $s$.

> The pattern $p$ should be the **first** element in $S$ in the ordering, and all other elements of $S$ must appear after $p$.

Reviewer 2: ```suggestion
The pattern $p$ should appear **before** all other patterns in $S$ in the ordering.
```

> Each requirement is therefore a **precedence constraint**, where $p$ cannot come before any other elements in $S$.

Reviewer 3: this contradicts to previous sentence if i am not wrong or should it be `p must come before ....`

Author: oops

Author: ty for spotting that

We model these as directed edges, $p \to x$, where $x$ is any non-$p$ element in $S$.

> A valid ordering then corresponds to a topological sort of the resulting graph: if the constraints are consistent, any topological order works; if two constraints contradict each other (e.g. $p \to x$ and $x \to p$), a cycle forms and no valid ordering exists.

Reviewer 3: shouldn't you explain here why `any topological order works`?

like if the order is consistent with all edges from p, then with every query p will be before any other match pattern which makes p to be first match

To compute $S$, note that since $k \le 4$, each query string has at most $2^k \le 16$ wildcard variants, obtained by replacing subsets of positions with `_`.
We store a hashmap from pattern string to index and look up each variant to collect all matching patterns efficiently.

For each query:

1. Generate $S$ by enumerating all $2^k$ wildcard variants.
2. If $p \notin S$, immediately output `NO`.
3. Otherwise, add directed edges $p \to x$ for all $x \in S$, $x \neq p$.

After processing all queries, run a topological sort on the pattern graph. If a cycle exists, print `NO`; otherwise, output the topological order.

## Implementation

> Since $k \le 4$, each query generates at most $2^k = 16$ variants, so building the graph takes $\mathcal{O}(m \cdot 2^k)$ time. The topological sort runs in

Reviewer 3: this line is already mentioned in the explanation. i don't think it is needed here
